x-pack/plugins/alerting/server/rules_client/methods/aggregate.ts

```diff
@@ -6,90 +6,34 @@
  */
 
 import { KueryNode, nodeBuilder } from '@kbn/es-query';
-import { RawRule, RuleExecutionStatusValues, RuleLastRunOutcomeValues } from '../../types';
+import type { AggregationsAggregationContainer } from '@elastic/elasticsearch/lib/api/typesWithBodyKey';
 import { AlertingAuthorizationEntity } from '../../authorization';
 import { ruleAuditEvent, RuleAuditAction } from '../common/audit_events';
 import { buildKueryNodeFilter } from '../common';
 import { alertingAuthorizationFilterOpts } from '../common/constants';
 import { RulesClientContext } from '../types';
+import { RawRule, AggregateOptions } from '../../types';
+import { validateRuleAggregationFields } from '../lib/validate_rule_aggregation_fields';
 
-export interface AggregateOptions extends IndexType {
-  search?: string;
-  defaultSearchOperator?: 'AND' | 'OR';
-  searchFields?: string[];
-  hasReference?: {
-    type: string;
-    id: string;
-  };
-  filter?: string | KueryNode;
-  maxTags?: number;
+export interface AggregateParams {
+  options?: AggregateOptions;
+  aggs: Record<string, AggregationsAggregationContainer>;
 }
 
-interface IndexType {
-  [key: string]: unknown;
-}
-
-export interface AggregateResult {
-  alertExecutionStatus: { [status: string]: number };
-  ruleLastRunOutcome: { [status: string]: number };
-  ruleEnabledStatus?: { enabled: number; disabled: number };
-  ruleMutedStatus?: { muted: number; unmuted: number };
-  ruleSnoozedStatus?: { snoozed: number };
-  ruleTags?: string[];
-}
-
-export interface RuleAggregation {
-  status: {
-    buckets: Array<{
-      key: string;
-      doc_count: number;
-    }>;
-  };
-  outcome: {
-    buckets: Array<{
-      key: string;
-      doc_count: number;
-    }>;
-  };
-  muted: {
-    buckets: Array<{
-      key: number;
-      key_as_string: string;
-      doc_count: number;
-    }>;
-  };
-  enabled: {
-    buckets: Array<{
-      key: number;
-      key_as_string: string;
-      doc_count: number;
-    }>;
-  };
-  snoozed: {
-    count: {
-      doc_count: number;
-    };
-  };
-  tags: {
-    buckets: Array<{
-      key: string;
-      doc_count: number;
-    }>;
-  };
-}
-
-export async function aggregate(
+export async function aggregate<T extends Record<string, unknown>>(
```

Comment:
Tbh `T extends Record<string, unknown>` creates inconvenience since provided type has to explicitly extend `Record<string, unknown>`. For example I wanna just to get numbers of prebult and custom rules so I need to declare the following type

```
interface RulesCountAggregationResult extends Record<string, unknown> {
  prebuiltRulesCount: {
    buckets: Array<{ doc_count: number }>;
  };
  customRulesCount: {
    buckets: Array<{ doc_count: number }>;
  };
}
``` 

while I have declared necessary fields I still need to extend the type from `Record<string, unknown>`. Basically it says the type can have arbitrary fields but it doesn't have them in fact. This way this can be relaxed by replacing `T extends Record<string, unknown>` with `T = Record<string, unknown>` or `T extends Record<string, any>`.

As another typing improvement it can request to define all output fields defined in the input I did [here](https://github.com/elastic/kibana/pull/148382/files#diff-ee172f284c3c756a3555fc0506cd23b77dae1df44c8d40e0bafd90e435ba73ffR37) by inferring them. 

SO client gives a lot of freedom in declaring return type though ES returns an explicit result. So rulesClient can fill this gap a bit.

Reply:
++ We should have a working type inference without the need to define types like this: `interface RulesCountAggregationResult extends Record<string, unknown>`

Reply:
Good point, I will make the suggested changes